rfcs/framework-0072-pluggable-transfers.md

```diff
@@ -0,0 +1,162 @@
+cases emerging
+
+- [ ] Clarify the "storage_class" vs "transfer_type"
+
+It seems that inside invenio_files the storage_class was meant to specify
+the details of the storage backend (as the tests suggest, at least) - for example
+'A' as 'Archive'.
+
+In invenio_records_resources, the storage_class on the initial schema means the type of 
+transfer that will be used (Local, Fetch, Remote). This is independent of the storage
+backend and is a bit confusing.
+
+- [ ] FileSchema.uri vs. links section
+
+For the initialization of files, 'uri' can be used to specify the location of the file
+for REMOTE/FETCH transfers. Later on, clients can use this URI for these transfers to get
+a direct access to the referenced data.
+
+During "dumping" the file record, a links section with `content` uri is created. 
+This is a bit confusing as clients do not know which URL should be used.
+
+Proposal: keep the uri only inside the initial file record. Do not serialize it
+afterwards. Clients should use the links section for all the URIs. The content
+url can return a HTTP 302 redirect to the actual file location. This way counters
+will work as well.
+
+- [ ] FETCH transfer type and private URIs
+
+The FETCH transfer type is used to fetch the file from a remote location. The uri
+of the external location is visible to anyone who has a read access to the record,
+not just the uploader. As the uri might contain credentials (e.g. S3 signed URL),
+this might be a security issue.
+
+Proposal: do not serialize the uri of the FETCH transfer type. the fetch_file task
+would have to be changed to not to use the service but fetch the uri directly.
+
+- [ ] IfFileIsLocal generators
+
+The permissions in https://github.com/inveniosoftware/invenio-records-resources/blob/master/invenio_records_resources/services/files/generators.py#L27
+take into account only the files with storage class (== transfer type) "L". This means that
+the permissions will fail for "M" (Multipart transport) files. Is this intended?
+
+The same problem is in invenio-rdm-records/services/generators.py
+
+Proposal: either change the meaning of "Locality" and keep the permissions as they are
+or change the permissions to take into account all the files. Example from RDM that I do 
+not understand (can_draft_set_content_files - why there is the can_review permission and 
+not record owner?):
+
+```python
+    can_draft_create_files = can_review
+    can_draft_set_content_files = [
+        # review is the same as create_files
+        IfFileIsLocal(then_=can_review, else_=[SystemProcess()])
+    ]
+    can_draft_get_content_files = [
+        # preview is same as read_files
+        IfFileIsLocal(then_=can_draft_read_files, else_=[SystemProcess()])
+    ]
+    can_draft_commit_files = [
+        # review is the same as create_files
+        IfFileIsLocal(then_=can_review, else_=[SystemProcess()])
+    ]
+```
+
+- [ ] MultipartStorageExt
```

Comment:
question: Invenio-Files-REST comes with some existing models for storing multipart-related info, like e.g. [`MultipartObject`](https://github.com/inveniosoftware/invenio-files-rest/blob/master/invenio_files_rest/models.py#L1524) and [`Part`](https://github.com/inveniosoftware/invenio-files-rest/blob/0b0cbafcc08dd2545117b318441757002dffd5ef/invenio_files_rest/models.py#L1737). It also implements the REST API for working with this, but it's very much bound to the Angular.js client-side library we picked at the time. For Zenodo (and all Invenio v3-based services at CERN) we also never used this implementation, but because of performance issues for multipart on our storage cluster at CERN.

I see that you've used `ObjectVersionTag`s for organizing the parts; was there a specific reason for not reusing the `MultipartObject`/`Part` classes? In any case, what matters is if it works, but I could also see it as an opportunity to consider purging all the multipart models from files-rest :)

Reply:
I've just extracted the global code to previously unused interface (going the "protocol" way), imho looks better.  https://github.com/mesemus/invenio-records-resources/blob/multipart-transfers/invenio_records_resources/services/files/transfer/providers/multipart.py